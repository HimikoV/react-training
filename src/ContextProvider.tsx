import { createContext, useEffect, useState, useMemo } from "react";
import { mapBasketData } from "./mappers/mapBasketData";
import { getDeliveryOptions } from "./service/getDeliveryOptions";
import { getOrderData } from "./service/getOrderData";
import { getPaymentOptions } from "./service/getPaymentOptions";
import { requestData } from "./service/requestData";
import { basketType } from "./types/basketType";
import { finalizeOptionsType } from "./types/finalizeOptionsType";
import { orderType } from "./types/orderType";
import { productType } from "./types/productType";
import { childrenInterface } from "./interfaces/childrenInterface";
import { basketProcessedType } from "./types/basketProcessedType";
import { contextProviderInterface } from "./interfaces/contextProviderInterface";

export const dataContext = createContext<contextProviderInterface>(
  {} as contextProviderInterface
);

const ContextProvider = ({ children }: childrenInterface) => {
  const [basketData, setBasketData] = useState<basketType[]>(
    JSON.parse(localStorage.getItem("basketData") || "[]")
  );
  const [filter, setFilter] = useState<string[]>([]);
  const [data, setData] = useState<productType[]>([]);
  const [loadLimit, setLoadLimit] = useState<number>(9);
  const [basketProcessedData, setBasketProcessedData] = useState<
    basketProcessedType[]
  >([]);
  const [deliveryOptions, setDeliveryOptions] = useState<finalizeOptionsType[]>(
    []
  );
  const [paymentOptions, setPaymentOptions] = useState<finalizeOptionsType[]>(
    []
  );
  const [open, setOpen] = useState<boolean>(false);
  const [orderData, setOrderData] = useState<orderType[]>([]);

  useEffect(() => {
    const fetchData = async () => {
      const responseData = await requestData();
      setData(responseData);
      const responseDelivery = await getDeliveryOptions();
      setDeliveryOptions(responseDelivery);
      const responsePayment = await getPaymentOptions();
      setPaymentOptions(responsePayment);
      const responseOrderData = await getOrderData();
      setOrderData(responseOrderData);
    };
    fetchData();
  }, []);

  useEffect(() => {
    localStorage.setItem("basketData", JSON.stringify(basketData));
  }, [basketData]);

  useEffect(() => {
    mapBasketData(data, basketData, setBasketProcessedData);
  }, [data]);

  const trueValues = useMemo(
    () => ({
      filter,
      setFilter,
      basketData,
      setBasketData,
      data,
      loadLimit,
      setLoadLimit,
      basketProcessedData,
      setBasketProcessedData,
      deliveryOptions,
      setDeliveryOptions,
      paymentOptions,
      setPaymentOptions,
      open,
      setOpen,
      orderData,
      setOrderData,
    }),
    [
      filter,
      setFilter,
      basketData,
      setBasketData,
      data,
      loadLimit,
      setLoadLimit,
      basketProcessedData,
      setBasketProcessedData,
      deliveryOptions,
      setDeliveryOptions,
      paymentOptions,
      setPaymentOptions,
      open,
      setOpen,
      orderData,
      setOrderData,
    ]
  );

  // const values = {};
  return (
    <dataContext.Provider value={trueValues}>{children}</dataContext.Provider>
  );
};
export default ContextProvider;
