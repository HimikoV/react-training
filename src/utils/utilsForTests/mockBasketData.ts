import { basketProcessedType } from "../../types/basketProcessedType";
import { basketType } from "../../types/basketType";

export const mockBasketData: basketType[] = [
  { id: "item_1", quantity: 4 },
  { id: "item_2", quantity: 5 },
  { id: "item_3", quantity: 1 },
];
export const mockBasketDataResult: basketType[] = [];
export const mockBasketProcessedData: basketProcessedType[] = [];
export const mockBasketProcessedDataResult: basketProcessedType[] = [
  {
    name: "product1",
    price: 15,
    category: "category1",
    id: "item_1",
    imgUrl: "url1",
    quantity: 4,
  },
  {
    name: "product2",
    price: 15,
    category: "category2",
    id: "item_2",
    imgUrl: "url2",
    quantity: 5,
  },
  {
    name: "product3",
    price: 15,
    category: "category3",
    id: "item_3",
    imgUrl: "url2",
    quantity: 1,
  },
];
