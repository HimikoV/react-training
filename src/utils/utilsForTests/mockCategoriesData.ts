import { categoryType } from "../../types/categoryType";

export const mockCategoriesData: categoryType[] = [
  {
    categoryName: "category1",
    categoryId: "category1",
  },
  {
    categoryName: "category2",
    categoryId: "category2",
  },
];
