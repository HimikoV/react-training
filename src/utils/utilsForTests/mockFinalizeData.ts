import { finalizeOptionsType } from "../../types/finalizeOptionsType";

export const mockFinalizeData: finalizeOptionsType[] = [
  {
    name: "DPD Courier",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: true,
  },
  {
    name: "DPD Inpost",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: false,
  },
];
export const mockFinalizeDataPayment: finalizeOptionsType[] = [
  {
    name: "Cash",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: true,
  },
  {
    name: "Card",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: false,
  },
];
export const mockFinalizeDataResult: finalizeOptionsType[] = [
  {
    name: "DPD Courier",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: true,
    key: "DPD Courier",
  },
  {
    name: "DPD Inpost",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: false,
    key: "DPD Inpost",
  },
];
