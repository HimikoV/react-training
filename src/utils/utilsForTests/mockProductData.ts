import { productType } from "../../types/productType";

export const mockProductData: productType[] = [
  {
    name: "product1",
    price: 15,
    category: "category1",
    id: "item_1",
    imgUrl: "url1",
  },
  {
    name: "product2",
    price: 15,
    category: "category2",
    id: "item_2",
    imgUrl: "url2",
  },
  {
    name: "product3",
    price: 15,
    category: "category3",
    id: "item_3",
    imgUrl: "url2",
  },
];
export const mockProductDataResult: productType[] = [
  {
    name: "product1",
    price: 15,
    category: "category1",
    id: "item_1",
    imgUrl: "url1",
    key: "item_1",
  },

  {
    name: "product2",
    price: 15,
    category: "category2",
    id: "item_2",
    imgUrl: "url2",
    key: "item_2",
  },
  {
    name: "product3",
    price: 15,
    category: "category3",
    id: "item_3",
    imgUrl: "url2",
    key: "item_3",
  },
];
