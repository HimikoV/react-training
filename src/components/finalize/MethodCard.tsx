import {
  Grid,
  CardContent,
  CardMedia,
  Typography,
  IconButton,
  Card,
} from "@mui/material";
import CheckBoxOutlinedIcon from "@mui/icons-material/CheckBoxOutlined";
import { useContext } from "react";
import { dataContext } from "../../ContextProvider";
import { methodCardInterface } from "../../interfaces/methodCardInterface";
import CheckBoxOutlineBlankOutlinedIcon from "@mui/icons-material/CheckBoxOutlineBlankOutlined";
import colors from "../../palette.module.scss";
import changeFinalizeOptions from "../../utils/componentsFunctions/changeFinalizeOptions";

const MethodCard = ({ name, imgUrl, selected }: methodCardInterface) => {
  const {
    deliveryOptions,
    setDeliveryOptions,
    paymentOptions,
    setPaymentOptions,
  } = useContext(dataContext);

  return (
    <Card
      sx={{
        display: "flex",
        flexWrap: "wrap",
        color: "#3d405b",
        justifyContent: "flex-start",
        alignItems: "center",
        flexDirection: "row",
        width: "100%",
        mb: "0.5rem",
        bgcolor: colors.colorAutumnLight50,
      }}
    >
      <Grid
        sx={{
          alignItems: "center",
          flexDirection: "row",
          width: "6.25rem",
        }}
      >
        <CardMedia
          component="img"
          height="100"
          image={imgUrl}
          alt="green iguana"
        />
      </Grid>

      <CardContent
        sx={{
          display: "flex",
          flexDirection: "row",
          justifyContent: "space-between",
          alignItems: "center",
          width: "11.25rem",
        }}
      >
        <Typography color="primary" variant="h5" sx={{ maxWidth: "9.375rem" }}>
          {name}
        </Typography>

        {selected ? (
          <IconButton
            onClick={() => {
              setDeliveryOptions(
                changeFinalizeOptions(deliveryOptions, paymentOptions, name)
                  .deliveryOptions
              );
              setPaymentOptions(
                changeFinalizeOptions(deliveryOptions, paymentOptions, name)
                  .paymentOptions
              );
            }}
          >
            <CheckBoxOutlinedIcon fontSize="medium" color="primary" />
          </IconButton>
        ) : (
          <IconButton
            onClick={() => {
              setDeliveryOptions(
                changeFinalizeOptions(deliveryOptions, paymentOptions, name)
                  .deliveryOptions
              );
              setPaymentOptions(
                changeFinalizeOptions(deliveryOptions, paymentOptions, name)
                  .paymentOptions
              );
            }}
          >
            <CheckBoxOutlineBlankOutlinedIcon
              fontSize="medium"
              color="primary"
            />
          </IconButton>
        )}
      </CardContent>
    </Card>
  );
};

export default MethodCard;
