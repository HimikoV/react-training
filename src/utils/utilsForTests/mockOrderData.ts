import { deliveryInformationType } from "../../types/deliveryInformationType";
import { orderType } from "../../types/orderType";
import { mockBasketData } from "./mockBasketData";

export const mockDeliveryInformation: deliveryInformationType = {
  name: "asdfasd",
  surname: "sdafas",
  email: "asdf@asdfjha.com",
  phone: "48325235235",
  zipcode: "23-325",
  state: "afdas",
  city: "asdfas",
};

export const mockOrderData: orderType[] = [
  {
    id: 1234,
    items: mockBasketData,
    date: "1/1/2022",
    delivery_information: mockDeliveryInformation,
    delivery_option: "DHL",
    payment_option: "cash",
  },
  {
    id: 1235,
    items: mockBasketData,
    date: "1/1/2022",
    delivery_information: mockDeliveryInformation,
    delivery_option: "DHL",
    payment_option: "cash",
  },
];
