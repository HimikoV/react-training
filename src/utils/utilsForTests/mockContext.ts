import {
  mockBasketData,
  mockBasketProcessedDataResult,
} from "./mockBasketData";
import { mockProductData, mockProductDataResult } from "./mockProductData";
import { mockFinalizeData } from "./mockFinalizeData";
import { mockOrderDataArray } from "./mockOrderData";
import { contextProviderInterface } from "../../interfaces/contextProviderInterface";
import { Firestore } from "firebase/firestore";
import { NULL } from "sass";
import { Auth } from "firebase/auth";

export const mockContext: contextProviderInterface = {
  filter: [],
  setFilter: jest.fn(),
  basketData: mockBasketData,
  setBasketData: jest.fn(),
  data: mockProductData,
  loadLimit: 9,
  setLoadLimit: jest.fn(),
  basketProcessedData: mockBasketProcessedDataResult,
  setBasketProcessedData: jest.fn(),
  deliveryOptions: mockFinalizeData,
  setDeliveryOptions: jest.fn(),
  paymentOptions: mockFinalizeData,
  setPaymentOptions: jest.fn(),
  open: false,
  setOpen: jest.fn(),
  orderData: mockOrderDataArray,
  setOrderData: jest.fn(),
  order: false,
  setOrder: jest.fn(),
  sort: "letter",
  setSort: jest.fn(),
  processedData: mockProductDataResult,
  setProcessedData: jest.fn(),
  orderLoadLimit: 20,
  setOrderLoadLimit: jest.fn(),
  firestore: {} as Firestore,
  auth: {} as Auth,
  user: null,
  setUser: jest.fn(),
};
