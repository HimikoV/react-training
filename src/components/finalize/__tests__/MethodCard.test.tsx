import renderer from "react-test-renderer";
import MethodCard from "../MethodCard";
import { dataContext } from "../../../ContextProvider";
import { mockContext } from "../../../utils/utilsForTests/mockContext";
import { mockFinalizeData } from "../../../utils/utilsForTests/mockFinalizeData";

describe("testing MethodCard render", () => {
  it("MethodCard Renders Correctly", () => {
    const methodCard = renderer
      .create(
        <dataContext.Provider value={mockContext}>
          <MethodCard
            name={mockFinalizeData[0].name}
            imgUrl={mockFinalizeData[0].imgUrl}
            selected={mockFinalizeData[0].selected}
          />
        </dataContext.Provider>
      )
      .toJSON();
    expect(methodCard).toMatchSnapshot();
  });
});
