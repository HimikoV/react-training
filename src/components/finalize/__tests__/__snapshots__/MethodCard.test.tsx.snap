// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing MethodCard render MethodCard Renders Correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-xbf7rn-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiGrid-root css-2etixe-MuiGrid-root"
  >
    <img
      alt="green iguana"
      className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
      height="100"
      src="https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80"
    />
  </div>
  <div
    className="MuiCardContent-root css-e1l2wi-MuiCardContent-root"
  >
    <h5
      className="MuiTypography-root MuiTypography-h5 css-kteq1r-MuiTypography-root"
    >
      DPD Courier
    </h5>
    <button
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
        data-testid="CheckBoxOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V5h14v14zM17.99 9l-1.41-1.42-6.59 6.59-2.58-2.57-1.42 1.41 4 3.99z"
        />
      </svg>
    </button>
  </div>
</div>
`;
